### PR TITLE
docs: add example for custom file format with `COPY TO`

### LINKED ISSUE #11079: Add example for writing a `FileFormat`
### Is your feature request related to a problem or challenge?

Now that @devinjdangelo has added better support for user defined file formats (see https://github.com/apache/datafusion/pull/11060) 

We have an example of a custom table provider: https://datafusion.apache.org/library-user-guide/custom-table-providers.html and example 

It would be great to have an example of creating a file format

Such an example would make sure the APIs added in this PR can be used for the end to end case, as well as obviously made it easier for others to discover and use this feature

### Describe the solution you'd like

Add an example in https://github.com/apache/datafusion/tree/main/datafusion-examples that does:
1. Creates a new file format (something basic like use CSV / ROT13)
2. reads / writes data using that format

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tshauck**
## Which issue does this PR close?

Closes #11079

## Rationale for this change

Adds an example of for how to `COPY table TO` a custom file format.

## What changes are included in this PR?

Created sort of a mock file format factory and file format that just wrap the CSV one but for a TSV.

## Are these changes tested?

yes, I've run the example

## Are there any user-facing changes?

no

### CODE REVIEW

**Comment by tshauck** [comment]
Seem to be failing in CI due to being out space on the CI worker.

image

image

File: ci/scripts/rust_example.sh [hunk 1]
```diff
@@ -29,5 +29,6 @@ do
   # Skip tests that rely on external storage and flight
   if [ ! -d $filename ]; then
      cargo run --example $example_name
+     cargo clean -p datafusion-examples
```
Comment (tshauck) [inline ci/scripts/rust_example.sh:32]: I can into consistent build issues where the runner would run out of disk space. This is the best solution I could come up with. It seems to add about 5-10 seconds to the example action, which takes about 12 minutes overall.

Certainly open to alternatives.

File: datafusion-examples/examples/custom_file_format.rs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+    let d = ctx
+        .sql(&format!(
+            "COPY mem_table TO '{}' STORED AS TSV;",
+            table_save_path.display(),
+        ))
+        .await?;
+
+    let results = d.collect().await?;
+    println!("Number of inserted rows: {:?}", results[0]);
```
Comment (devinjdangelo) [inline datafusion-examples/examples/custom_file_format.rs:201]: We could make the print a bit cleaner with something like:

```suggestion
println!(
        "Number of inserted rows: {:?}",
        (results[0]
            .column_by_name("count")
            .unwrap()
            .as_primitive::<UInt64Type>()
            .value(0))
    );
```

**Review by devinjdangelo — APPROVED** [review]
Thank you @tshauck for putting this example together! I think that wrapping an existing `FileFormat` was a clever way to demonstrate this without tons of boilerplate code being required for a working example. It ran locally for me no issues.

I left just one small suggestion.

File: datafusion-examples/examples/custom_file_format.rs (hunk 2 above)
Comment (tshauck) [reply datafusion-examples/examples/custom_file_format.rs]: Thanks! Done in https://github.com/apache/datafusion/pull/11174/commits/77ff942ac635499d42cb5ac3c21dca214bb7e6db

File: ci/scripts/rust_example.sh (hunk 1 above)
Comment (alamb) [reply ci/scripts/rust_example.sh:32]: I am hitting similar issues with trying to add new examples in https://github.com/apache/datafusion/pull/11089

This seems like a good idea to me

My best hope is to move some of the example binaries into inline examples in the docs instead: https://github.com/apache/datafusion/pull/11178

Hopefully that will free up some additional space as well as make the examples easier to navigate

**Review by alamb — APPROVED** [review]
Thank you so much @tshauck  -- I also think this looks really nice. 

Thank you for the review @devinjdangelo 

I agree the idea of wrapping an existing format to show the API is clever.  I was thinking it would be awesome to somehow cook up a simple to implement custom format that would handle things entirely end to end but I can't think of any format suitable simple.

**Comment by tshauck** [comment]
Yeah, I started down the road of something custom, but it turned out to be a non-trivial amount of code.

Perhaps adding a link to something like https://github.com/datafusion-contrib/datafusion-orc would show a more complex example, though it doesn't implement `create_writer_physical_plan` yet.

**Comment by alamb** [comment]
Thanks again @tshauck

File: ci/scripts/rust_example.sh (hunk 1 above)
Comment (tshauck) [reply ci/scripts/rust_example.sh:32]: er, sorry this came up. There are a few other tricks here (https://github.com/actions/runner-images/issues/2840#issuecomment-790492173) to clear up some disk space that may help again.

Comment (alamb) [reply ci/scripts/rust_example.sh:32]: No worries! I think this is a good excuse / reason to take another pass through the examples directory (and the library guide that you started however long ago).

**Pull request merged**